app/web_framework/cluster_master_application.py

```diff
@@ -253,3 +255,15 @@ def get(self):
         self.write({
             'events': analytics.get_events(since_timestamp, since_id),
         })
+
+
+class _SlaveShutdownHandler(_ClusterMasterBaseHandler):
+    @authenticated
+    def post(self):
+        shutdown_all = self.decoded_body.get('shutdown_all')
+        slaves_to_shutdown = self._cluster_master.all_slaves_by_id().keys() if shutdown_all else\
+            [int(slave_id) for slave_id in self.decoded_body.get('slaves')]
+
+        for slave_id in slaves_to_shutdown:
+            slave = self._cluster_master.get_slave(slave_id)
```

Comment:
@tjlee0909 are you suggesting that it *should* return a 500? My point was that currently as it is written, the code would return a 404 since the `self._cluster_master.get_slave(slave_id)` call would raise an `ItemNotFoundError` which will be translated to a 404. My point was that when that exception is raised, we'll already have shut down some slaves, so it's a bit surprising.

The alternative would be to build up the list of slaves first (before calling `handle_slave_state_update(slave, SlaveState.SHUTDOWN)` on any of them). While we're checking this, maybe we'd also want to convert that 404 to something else (maybe just a 400?). Then once we're sure all the provided IDs are valid slaves, that's when iterate through them and tell each to shut down.